firmware-support/bittide-sys/src/time.rs

```diff
@@ -0,0 +1,330 @@
+
+    fn add(self, rhs: Duration) -> Duration {
+        Duration {
+            micros: self.micros + rhs.micros,
+        }
+    }
+}
+
+impl ops::AddAssign<Duration> for Duration {
+    fn add_assign(&mut self, rhs: Duration) {
+        self.micros += rhs.micros;
+    }
+}
+
+impl ops::Sub<Duration> for Duration {
+    type Output = Duration;
+
+    fn sub(self, rhs: Duration) -> Duration {
+        Duration {
+            micros: self
+                .micros
+                .checked_sub(rhs.micros)
+                .expect("overflow when subtracting durations"),
+        }
+    }
+}
+
+/// Class to utilize an overflowing free-running counter as clock.
+/// Allows us to measure time as `Instant`s and wait for certain `Duration`s.
+#[derive(uDebug)]
+pub struct Clock {
+    global_cycles: Instant,
+    counter_last: usize,
+    counter_now: *const usize,
+}
+
+impl Clock {
+    /// Create a new Clock instance.
+    ///
+    /// # Safety
+    ///
+    /// `addr` needs to point to a mapped memory address for a timer component.
+    pub fn new(addr: usize) -> Clock {
+        let counter_last = unsafe { (addr as *const usize).read_volatile() };
+        let frequency = unsafe { ((4 + addr) as *const u64).read_volatile() };
+        Clock {
+            global_cycles: Instant::from_cycles(counter_last as u64, frequency),
+            counter_last,
+            counter_now: addr as *mut usize,
+        }
+    }
+
+    /// Wait for a `Duration`.
+    pub fn wait(&mut self, duration: &Duration) {
+        let mut now = self.elapsed();
+        let target = now + *duration;
+        while target > now {
+            now = self.elapsed();
+        }
+    }
+
+    /// Wait until we have passed an `Instant`.
+    pub fn wait_until(&mut self, target: Instant) {
+        let mut now = self.elapsed();
+        while target > now {
+            now = self.elapsed();
+        }
+    }
+
+    /// Update the `Clock`, if the `Clock` is not regularly used, it's important that
+    /// this function is periodically called to handle overflows of the peripheral's counter.
+    pub fn update(&mut self) {
+        let old = self.counter_last;
+        let new = self.get_ticks();
+        let diff_ticks: usize = if old > new { !0 - old + new } else { new - old };
+        self.global_cycles.clock_cycles += diff_ticks as u64;
+        self.counter_last = new;
+    }
+
+    /// Update the clock and return the current Instant.
+    pub fn elapsed(&mut self) -> Instant {
+        self.update();
+        self.global_cycles
+    }
+
+    /// Get the last read value of the free running counter.
+    pub fn get_counter_last(&mut self) -> usize {
+        self.counter_last
+    }
+
+    pub fn get_frequency(&mut self) -> u64 {
+        self.global_cycles.get_frequency()
+    }
+    /// Gets the current current number of ticks in usize, this will overflow during runtime.
+    fn get_ticks(&self) -> usize {
+        unsafe { self.counter_now.read_volatile() }
+    }
```

Comment:
I'm not having a good feeling about this, because the documentation of the Wishbone component says that overflows needs to be accounted for and it seems like that's not happening here anywhere!

Would it make sense to have something like `get_ticks_since_last_check()` that could then try to recover the number of ticks even across (one) overflow? If we need this timer component to also be usable in longer running applications we need to think about how we want to deal with overflows and representing time in general.

We can keep track of changes of number of clock cycles as long as it's not more than one "overflow space", but we might want to refer to time with `Instant` or `Duration` that's longer than that? We can use `u64` or `u128` just fine, it will just be slower.

Just something to think about because right now we're not doing anything :see_no_evil:

Reply:
I discussed this privately, the overflow gets handled in `update()` but I'll remove this function and relocate the functionality such that it's paired with the overflow handling.